fvm/fvm_test.go

```diff
@@ -1893,3 +1893,208 @@ func TestSigningWithTags(t *testing.T) {
+			fundWith:      txFees,
+			tryToTransfer: 1,
+			checkResult: func(t *testing.T, balanceBefore uint64, balanceAfter uint64, tx *fvm.TransactionProcedure) {
+				require.IsType(t, tx.Err, &errors.TransactionFeeDeductionFailedError{})
+				require.Equal(t, txFees, balanceAfter)
+			},
+		},
+		{
+			// This will change later
+			name:          "If tx fails, no fees are deducted",
+			fundWith:      fundingAmount,
+			tryToTransfer: 2 * fundingAmount,
+			checkResult: func(t *testing.T, balanceBefore uint64, balanceAfter uint64, tx *fvm.TransactionProcedure) {
+				require.Equal(t, fundingAmount, balanceAfter)
+			},
+		},
+	}
+
+	for i, tc := range testCases {
+
+		t.Run(fmt.Sprintf("Transaction Fees %d: %s", i, tc.name), newVMTest().withBootstrapProcedureOptions(
+			fvm.WithTransactionFee(fvm.DefaultTransactionFees),
+		).run(
+			func(t *testing.T, vm *fvm.VirtualMachine, chain flow.Chain, ctx fvm.Context, view state.View, programs *programs.Programs) {
+				ctx.TransactionFeesEnabled = true
+				// Create an account private key.
+				privateKeys, err := testutil.GenerateAccountPrivateKeys(1)
+				require.NoError(t, err)
+
+				// Bootstrap a ledger, creating accounts with the provided private keys and the root account.
+				accounts, err := testutil.CreateAccounts(vm, view, programs, privateKeys, chain)
+				require.NoError(t, err)
+
+				// ==== Transfer tokens to new account ====
+
+				txBody := transferTokensTx(chain).
+					AddAuthorizer(chain.ServiceAddress()).
+					AddArgument(jsoncdc.MustEncode(cadence.UFix64(tc.fundWith))).
+					AddArgument(jsoncdc.MustEncode(cadence.NewAddress(accounts[0])))
+
+				txBody.SetProposalKey(chain.ServiceAddress(), 0, 0)
+				txBody.SetPayer(chain.ServiceAddress())
+
+				err = testutil.SignPayload(
+					txBody,
+					chain.ServiceAddress(),
+					unittest.ServiceAccountPrivateKey,
+				)
+				require.NoError(t, err)
+
+				err = testutil.SignEnvelope(
+					txBody,
+					chain.ServiceAddress(),
+					unittest.ServiceAccountPrivateKey,
+				)
+				require.NoError(t, err)
+
+				tx := fvm.Transaction(txBody, 0)
+
+				err = vm.Run(ctx, tx, view, programs)
+				require.NoError(t, err)
+				require.NoError(t, tx.Err)
+
+				balanceBefore := getBalance(vm, chain, ctx, view, accounts[0])
+
+				// ==== Transfer tokens from new account ====
+
+				txBody = transferTokensTx(chain).
+					AddAuthorizer(accounts[0]).
+					AddArgument(jsoncdc.MustEncode(cadence.UFix64(tc.tryToTransfer))).
+					AddArgument(jsoncdc.MustEncode(cadence.NewAddress(chain.ServiceAddress())))
+
+				txBody.SetProposalKey(accounts[0], 0, 0)
+				txBody.SetPayer(accounts[0])
+
+				err = testutil.SignPayload(
+					txBody,
+					accounts[0],
+					privateKeys[0],
+				)
+				require.NoError(t, err)
+
+				err = testutil.SignEnvelope(
+					txBody,
+					accounts[0],
+					privateKeys[0],
+				)
+				require.NoError(t, err)
+
+				tx = fvm.Transaction(txBody, 1)
+
+				err = vm.Run(ctx, tx, view, programs)
+				require.NoError(t, err)
```

Comment:
could you please add extra checks here to check if token transfer events are captured when fees are deducted.

Reply:
this test case
```
name:          "Transaction fee deduction emits events",
```

already does that. Unless you had something more specific in mind.

Reply:
ah I see, somehow I missed it. sorry